controllers/core/openstackclient_controller.go

```diff
@@ -0,0 +1,288 @@
+	_ = r.Log.WithValues("openstackclient", req.NamespacedName)
+
+	instance := &corev1beta1.OpenStackClient{}
+	err := r.Client.Get(context.TODO(), req.NamespacedName, instance)
+	if err != nil {
+		if k8s_errors.IsNotFound(err) {
+			r.Log.Info("OpenStackClient CR not found", "Name", instance.Name, "Namespace", instance.Namespace)
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+	r.Log.Info("OpenStackClient CR values", "Name", instance.Name, "Namespace", instance.Namespace, "Secret", instance.Spec.OpenStackConfigSecret, "Image", instance.Spec.ContainerImage)
+
+	instance.Status.Conditions = condition.Conditions{}
+	cl := condition.CreateList(
+		condition.UnknownCondition(
+			corev1beta1.OpenStackClientReadyCondition,
+			condition.InitReason,
+			corev1beta1.OpenStackClientReadyInitMessage,
+		),
+	)
+	instance.Status.Conditions.Init(&cl)
+
+	h, err := helper.NewHelper(
+		instance,
+		r.Client,
+		r.Kclient,
+		r.Scheme,
+		r.Log,
+	)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	// Always patch the instance status when exiting this function so we can persist any changes.
+	defer func() {
+		// update the overall status condition if service is ready
+		if instance.Status.Conditions.IsTrue(corev1beta1.OpenStackClientReadyCondition) {
+			instance.Status.Conditions.MarkTrue(condition.ReadyCondition, condition.ReadyMessage)
+		}
+
+		if err := h.SetAfter(instance); err != nil {
+			util.LogErrorForObject(h, err, "Set after and calc patch/diff", instance)
+		}
+
+		if changed := h.GetChanges()["status"]; changed {
+			patch := client.MergeFrom(h.GetBeforeObject())
+
+			if err := r.Status().Patch(ctx, instance, patch); err != nil && !k8s_errors.IsNotFound(err) {
+				util.LogErrorForObject(h, err, "Update status", instance)
+			}
+		}
+	}()
+
+	//
+	// Validate that keystoneAPI is up
+	//
+	keystoneAPI, err := keystonev1.GetKeystoneAPI(ctx, h, instance.Namespace, map[string]string{})
+	if err != nil {
+		if k8s_errors.IsNotFound(err) {
+			instance.Status.Conditions.Set(condition.FalseCondition(
+				keystonev1.KeystoneAPIReadyCondition,
+				condition.ErrorReason,
+				condition.SeverityWarning,
+				corev1beta1.OpenStackClientReadyErrorMessage,
+				err.Error()))
+			r.Log.Info("KeystoneAPI not found!")
+			return ctrl.Result{RequeueAfter: time.Second * 5}, nil
+		}
+		instance.Status.Conditions.Set(condition.FalseCondition(
+			corev1beta1.OpenStackClientReadyCondition,
+			condition.ErrorReason,
+			condition.SeverityWarning,
+			corev1beta1.OpenStackClientReadyErrorMessage,
+			err.Error()))
+		return ctrl.Result{}, err
+	}
+	if !keystoneAPI.IsReady() {
+		instance.Status.Conditions.Set(condition.FalseCondition(
+			corev1beta1.OpenStackClientReadyCondition,
+			condition.RequestedReason,
+			condition.SeverityInfo,
+			corev1beta1.OpenStackClientKeystoneWaitingMessage))
+		r.Log.Info("KeystoneAPI not yet ready")
+		return ctrl.Result{RequeueAfter: time.Second * 5}, nil
```

Comment:
Here we could watch KeystoneAPI instead. But this is not super important.

Reply:
I did not do that in other operators, where we check if keystoneAPI is up as then we'd reconcile on every KeystoneAPI change. I guess we should only do that if that is required/wanted.

Reply:
It is a trade off. If the data we use from the KeystoneAPI could change (i.e. the endpoint URLs change) then reconciling on that change is a good thing.

Reply:
I suggest we not watch KeystoneAPI for changes. It might kill the OpenStackClient pod and potentially disrupt someone logged in there if it changes. If there is a keystoneAPI update I think we could probably have a status that the OpenStackclient pod is stale and may need to be re-created. This would only be required if the password or something changed right?

Reply:
Yes a password change is a good example where the client CR needs to be reconciled. In this case it does not matter that there is a logged in user as if the keystone password is changed then that user cannot use the client any more :) 

Anyhow as I said it is not super important right now.